### PR TITLE
Support max_tasks_per_stage for scan

### PR DESCRIPTION
**Pull request by cemcayiroglu**
stage.max-tasks-per-stage configuration property can be used by to limit the number of tasks for scan.  

== RELEASE NOTES ==

General Changes

- Respect stage.max-tasks-per-stage to limit number of tasks for scan.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java [hunk 1]
```diff
@@ -44,7 +44,12 @@
      * If we cannot find an assignment for a split, it is not included in the map. Also returns a future indicating when
      * to reattempt scheduling of this batch of splits, if some of them could not be scheduled.
      */
-    SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks);
+    SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, int limit);
+
+    default SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: This is used only by tests. Could you please migrate the tests instead of leaving it in the interface?

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java [hunk 2]
```diff
@@ -99,13 +99,14 @@ public InternalNode selectCurrentNode()
...
     @Override
-    public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks)
+    public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, int limit)
     {
         Multimap<InternalNode, Split> assignment = HashMultimap.create();
         NodeMap nodeMap = this.nodeMap.get().get();
         NodeAssignmentStats assignmentStats = new NodeAssignmentStats(nodeTaskMap, nodeMap, existingTasks);
 
-        ResettableRandomizedIterator<InternalNode> randomCandidates = randomizedNodes(nodeMap, includeCoordinator, ImmutableSet.of());
+        ResettableRandomizedIterator<InternalNode> randomCandidates =
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java]: I might be missing something, but I'm not sure i understand how it works.

1. `NodeSelector#computeAssignments` is called multiple times as new splits are scheduled
2. `randomCandidates` is generated every time the `NodeSelector#computeAssignments` is called

Although a single `randomCandidates` size is less then `limit`, it doesn't guarantee same set of nodes to be selected. As `NodeSelector#computeAssignments` is celled again and again - new nodes are likely to be pulled in for scheduling, by effectively eventually returning all the nodes.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/TopologyAwareNodeSelector.java [hunk 3]
```diff
@@ -113,7 +113,7 @@ public InternalNode selectCurrentNode()
     }
 
     @Override
-    public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks)
+    public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, int limit)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/TopologyAwareNodeSelector.java]: Why doesn't the `TopologyAwareNodeSelector` respect the limit?

File: presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java [hunk 4]
```diff
@@ -424,6 +424,24 @@ public void testSplitCount()
...
+        ImmutableList.Builder<Split> initialSplits = ImmutableList.builder();
+        for (int i = 0; i < 10; i++) {
+            initialSplits.add(new Split(CONNECTOR_ID, transactionHandle, new TestSplitRemote()));
+        }
+
+        Set<Split> splits = new HashSet<>();
+        for (int i = 0; i < 5; i++) {
+            splits.add(new Split(CONNECTOR_ID, transactionHandle, new TestSplitRemote()));
+        }
+        Multimap<InternalNode, Split> assignments = nodeSelector.computeAssignments(splits, ImmutableList.copyOf(taskMap.values()), 2).getAssignments();
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java]: I think this test has to be changed to call the `computeAssignments` multiple times (same as it is done for real query scheduling)

Comment (arhimondr) [reply presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java]: Also please make sure the solution works in real deployment. As the scheduling code might get quite tricky.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/TopologyAwareNodeSelector.java (hunk 3 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/TopologyAwareNodeSelector.java]: I am not sure if it makes sense to use limit here. The topology awere selector selects the best nodes for splits for maximize locality. I am not sure we can even apply a limit here.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java (hunk 2 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java]: You are right! I will use a local variable to handle this case.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java (hunk 1 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: wil do.

File: presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java (hunk 4 above)
Comment (cemcayiroglu) [reply presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java]: will do.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java [hunk 5]
```diff
@@ -55,6 +55,8 @@
     private final int maxSplitsPerNode;
     private final int maxPendingSplitsPerTask;
 
+    private final AtomicReference<ResettableRandomizedIterator<InternalNode>> randomCandidates = new AtomicReference<>(null);
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java]: A `ResettableRandomizedIterator` is basically just a `List` of the originally data: https://github.com/prestodb/presto/blob/2e29401e90ab444a94ed9f3a836c6d980928dec3/presto-main/src/main/java/com/facebook/presto/execution/scheduler/ResettableRandomizedIterator.java#L23-L27

Should we just keep a `List` here? In that case we can just use `NodeSelector#selectRandomNodes ` . @arhimondr ? 

I remember this discussion happens before, let me see if I can find the previous discussion.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/TopologyAwareNodeSelector.java (hunk 3 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/TopologyAwareNodeSelector.java]: @cemcayiroglu : Note even for `SimpleNodeSelector`, when the splits doesn't support remote access (e..g Raptor), the limit is not applied there. cc @highker 

On the other hand, for `TopologyAwareNodeSelector`, the current algorithm makes best effort to co-locate data and compute nodes. Thus we should still be able to apply a limit over the candidate nodes and do best-effort topology-aware selection. Does that make sense?

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java [hunk 6]
```diff
@@ -44,7 +44,7 @@
      * If we cannot find an assignment for a split, it is not included in the map. Also returns a future indicating when
      * to reattempt scheduling of this batch of splits, if some of them could not be scheduled.
      */
-    SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks);
+    SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, int limit);
```
Comment (wenleix) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: Instead of making `limit` a parameter to `computeAssignments`, another idea is to allow construct `NodeSelector` over a random set of nodes?  For example, today we construct `NodeSelector` through:

```java
            NodeSelector nodeSelector = nodeScheduler.createNodeSelector(connectorId);
```

We can have a different API: 
```java
            NodeSelector nodeSelector = nodeScheduler.createNodeSelector(connectorId, maxNodeCount);
```

I am suggesting this because today, once `NodeSelector#computeAssignments` is called, it memorize 
`randomCandidates` based on `limit`. It makes `NodeSelector` stateful which is a bit difficult to reason the behavior (e.g. what about the `NodeSelector` get called in other place) 

What do you think? @highker , @arhimondr ?

Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: I think @wenleix's suggestion is much better. Having `limit` on the interface is not intuitive to understand

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/TopologyAwareNodeSelector.java (hunk 3 above)
Comment (highker) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/TopologyAwareNodeSelector.java]: Agree, let's hide `limit` from those non-applicable cases.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java (hunk 6 above)
Comment (aweisberg) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: I think the direction Cem is heading in is that node selection won't be random it will be based on resources. So constructing node selector with a subset of the nodes would mean it could no longer make decisions resource based decisions.

Even today we have a TopologyAwareNodeSelector and if we passed that a random set of nodes at construction then it might be restricted as to what topology aware scheduling it can do.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java [hunk 7]
```diff
@@ -168,6 +170,14 @@ else if (!splitWaitingForAnyNode) {
         return new SplitPlacementResult(blocked, assignment);
     }
 
+    private ResettableRandomizedIterator<InternalNode> getRandomCandidates(int limit, NodeMap nodeMap)
+    {
+        if (randomCandidates.get() == null) {
+            randomCandidates.compareAndSet(null, new ResettableRandomizedIterator<>(selectNodes(limit, randomizedNodes(nodeMap, includeCoordinator, ImmutableSet.of()))));
```
Comment (aweisberg) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java]: It's not intuitive that this returns the same list if it is already constructed regardless of the parameters. Is this stuff all done exactly once? If it's done multiple times why is it guaranteed parameters like limit haven't changed?

If this needs to be the case then should it error if the parameters have changed?

File: presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java [hunk 8]
```diff
@@ -424,6 +425,28 @@ public void testSplitCount()
...
+        assertEquals(assignments.keySet().size(), 2);
+        Multimap<InternalNode, Split> tmp = assignments;
+        assignments = nodeSelector.computeAssignments(splits, ImmutableList.copyOf(taskMap.values()), 2).getAssignments();
+        // second call should return same nodes
```
Comment (aweisberg) [inline presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java]: What behavior are we testing here where the same set of nodes is returned each time? When would we want to get the list of nodes multiple times deterministically?

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java (hunk 6 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: @wenleix @highker @aweisberg Thanks for the input. I agree all of you. I had a trouble to weher to place the limit. Actually, there is a bigger problem here that creates confusion. DynamicSplitPlacementPolicy should define the node selection criteria like is remote or limited etc... Today it does almost nothing. Delegates everything to nodeselector. BucketedSplitPlacementPolicy is very similar. I think nodeselector should select nodes based on the criteria provided by SplitPlacementPolicy.  As @aweisberg said we are going to have different policies in the future to select nodes for memory aware scheduling. I am thinking of following what @wenleix described and hide limit. I will create a follow up PR for refactoring this parts to get ready for memory aware scheduling. Ant thoughts?

Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: After a second thought I will pass session to createNodeSelector instead of limit. nodeScheduler will access to limit if it is needed. Cleaner this way.

`NodeSelector nodeSelector = nodeScheduler.createNodeSelector(connectorId, session);`

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: Let me put my 5 cents into this discussion:

I would like to suggest some changes to the interfaces:

Currently the `NodeSelector` has 2 methods:
   - `SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, BucketNodeMap bucketNodeMap)`
  - `SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks)`

The first method accepts the `BucketNodeMap bucketNodeMap`. The `BucketNodeMap` has information about the selected nodes. The responsibility of selecting the nodes is not on that method. The node selection is happening outside this method. 

Thus, it sounds natural for the other method with exactly same name (that does random non bucketed assignments) to also have only the responsibility of assigning splits (randomly?), and not selecting the nodes. 

I would like to suggest to change the second method signature to something like `SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTask> existingTasks, List<Node> selectedNodes)`

As a "wish" (but not must) i may also suggest to rename the first method into something like `computeBucketedAssignments`, and the second is `computeUnbucketedAssignments`. Or something like that.

The responsibility of selecting the nodes should be moved to the `DynamicSplitPlacementPolicy` then (or even outside of this class, to be consistent with how it is done for the  `BucketedSplitPlacementPolicy`.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: Le me know how does it sound

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: The other important question is that node selection is currently happening continuously. By memoizing the selected node we are breaking the current semantic. I'm not sure how important is that. Maybe it is not.

But, if we really want to keep the semantics - in the `DynamicSplitPlacementPolicy` the nodes have to be selected every time. But it should also keep track of the nodes that already have some splits assigned. This list should be extended every time a new node is selected. And once the size limit is reached - the extention might be terminted, and only the already selected nodes be scheduled.

Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: @arhimondr this sounds great! Do you think it is better to handle the refactoring part on another PR? you are also right about the comment of the semantics. The current way can cause a problem if the initial split size is lower than the limit and the next call can be over the limit. Enough nodes will not be selected. I am going to make this change or find another way to avoid this.

Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeSelector.java]: @arhimondr  Actually we need to keep as is if we want to keep to semantics same. The current code does not support node addition. Nodemap is created by "createNodeSelector " and stays the same throughout the query execution. If it is ok I would like to merge it as is.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java [hunk 9]
```diff
@@ -105,7 +110,7 @@ public SplitPlacementResult computeAssignments(Set<Split> splits, List<RemoteTas
         NodeMap nodeMap = this.nodeMap.get().get();
         NodeAssignmentStats assignmentStats = new NodeAssignmentStats(nodeTaskMap, nodeMap, existingTasks);
 
-        ResettableRandomizedIterator<InternalNode> randomCandidates = randomizedNodes(nodeMap, includeCoordinator, ImmutableSet.of());
+        ResettableRandomizedIterator<InternalNode> randomCandidates = getRandomCandidates(maxTasksPerStage, nodeMap);
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java]: I still don't like the idea with memoizing. The `NodeMap` right now is mutable (although with 5 second memoization). Fixing the selected nodes will change the existing semantics and cluster behaviour. Although i don't know how big a problem is that, i would still prefer to keep behaviour the same.

The `List<RemoteTask> existingTasks` is the list of existing tasks.

`RemoteTask#getNodeId` returns the assigned node id of the task. To keep the existing semantins we should probably change the algorithm to something like:

* Get all the node
* Filter out all the nodes that already have tasks
* Select random N nodes from the remaining nodes, where N  = maxTasksPerStage - existingTasks.size()
* Union nodes with existing tasks ...

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 10]
```diff
@@ -421,7 +422,7 @@ private static URI getBufferLocation(RemoteTask remoteTask, OutputBufferId rootB
             if (isInternalSystemConnector(connectorId)) {
                 connectorId = null;
             }
-            NodeSelector nodeSelector = nodeScheduler.createNodeSelector(connectorId);
+            NodeSelector nodeSelector = nodeScheduler.createNodeSelector(connectorId, maxTasksPerStage);
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: The interface now is really weird.

`nodeScheduler.createNodeSelector` now acceps the `maxTasksPerStage`. And then the `nodeSelector.selectRandomNodes` accepts the `maxTasksPerStage`

I wonder if passing the limit to the `NodeSelector#computeAssignments` is a lesser evil?

CC: @wenleix

File: presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java [hunk 11]
```diff
@@ -170,7 +170,8 @@ public NetworkLocation get(HostAddress host)
             nonRackLocalBuilder.add(new Split(CONNECTOR_ID, transactionHandle, new TestSplitRemote(HostAddress.fromParts("data.other_rack", 1))));
         }
         Set<Split> nonRackLocalSplits = nonRackLocalBuilder.build();
-        Multimap<InternalNode, Split> assignments = nodeSelector.computeAssignments(nonRackLocalSplits, ImmutableList.copyOf(taskMap.values())).getAssignments();
+        Multimap<InternalNode, Split> assignments =
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java]: nit: unrelated change, please revert

File: presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java (hunk 8 above)
Comment (arhimondr) [reply presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java]: Actually yes, we want to test that if `computeAssignments` is called multiple times, no more than the nodes limit is assigned.

File: presto-main/src/test/java/com/facebook/presto/execution/scheduler/TestSourcePartitionedScheduler.java [hunk 12]
```diff
@@ -445,7 +445,8 @@ private static StageScheduler getSourcePartitionedScheduler(
                 .setMaxPendingSplitsPerTask(0);
         NodeScheduler nodeScheduler = new NodeScheduler(new LegacyNetworkTopology(), nodeManager, nodeSchedulerConfig, nodeTaskMap);
         SplitSource splitSource = new ConnectorAwareSplitSource(CONNECTOR_ID, TestingTransactionHandle.create(), connectorSplitSource);
-        SplitPlacementPolicy placementPolicy = new DynamicSplitPlacementPolicy(nodeScheduler.createNodeSelector(splitSource.getConnectorId()), stage::getAllTasks);
+        SplitPlacementPolicy placementPolicy =
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/execution/scheduler/TestSourcePartitionedScheduler.java]: nit: unrelated change, please revert

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java (hunk 10 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: I think changing `NodeSelector#computeAssignments`is a more evil cuz it will require to change TopologyAwareNodeSelector too.

Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: > I think changing NodeSelector#computeAssignmentsis a more evil

Let me elaborate a little bit more on this:

Creating the `NodeSelector` with the limit (that in fact the `maxTasksPerStage` is) is quite confusing.

The `NodeSelector` has other methods, that accept the limit as parameters. Thus it is not clear what limit will be applied, say in the `NodeSelector#selectRandomNodes`? The `maxTasksPerStage` limit? Or the `selectRandomNodes(limit..`)? Or `min()` / `max()`?

> it will require to change TopologyAwareNodeSelector too.

Currently it is simply ignored anyway. So i don't see a big difference.

Anyhow, i don't feel particularly strong about this. Feel free to leave it as is if you do.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeMap.java [hunk 13]
```diff
@@ -18,26 +18,35 @@
...
 public class NodeMap
 {
+    private final Map<String, InternalNode> nodesByNodeId;
     private final SetMultimap<HostAddress, InternalNode> nodesByHostAndPort;
     private final SetMultimap<InetAddress, InternalNode> nodesByHost;
     private final SetMultimap<NetworkLocation, InternalNode> workersByNetworkPath;
     private final Set<String> coordinatorNodeIds;
 
-    public NodeMap(SetMultimap<HostAddress, InternalNode> nodesByHostAndPort,
+    public NodeMap(Map<String, InternalNode> nodesByNodeId,
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeMap.java]: nit: new line (wrap it)

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java [hunk 14]
```diff
@@ -168,6 +171,19 @@ else if (!splitWaitingForAnyNode) {
         return new SplitPlacementResult(blocked, assignment);
     }
 
+    private ResettableRandomizedIterator<InternalNode> getRandomCandidates(int limit, NodeMap nodeMap, List<RemoteTask> existingTasks)
+    {
+        Set<InternalNode> existingNodes = existingTasks.stream().map(remoteTask -> nodeMap.getNodesByNodeId().get(remoteTask.getNodeId())).collect(toSet());
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java]: suggestion:

This code is gonna be called for every batch of splits. We can make it more efficient by

1. Collecting it here to a `List`
2. Create a set only in the less "likely" branch: `alreadySelectedNodeCount < limit`, right before it has to be passed to the `selectNodes`

File: presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java [hunk 15]
```diff
@@ -424,6 +424,47 @@ public void testSplitCount()
...
+            MockRemoteTaskFactory.MockRemoteTask remoteTask = remoteTaskFactory.createTableScanTask(taskId, node, ImmutableList.copyOf(assignments.get(node)), nodeTaskMap.createPartitionedSplitCountTracker(node, taskId));
+            remoteTask.startSplits(25);
+            nodeTaskMap.addTask(node, remoteTask);
+            taskMap.put(node, remoteTask);
+        }
+        assertEquals(assignments.keySet().size(), 2);
+        Multimap<InternalNode, Split> tmp = assignments;
+        assignments = nodeSelector.computeAssignments(splits, ImmutableList.copyOf(taskMap.values())).getAssignments();
+        // second call should return same nodes
+        assertEquals(assignments.keySet(), tmp.keySet());
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java]: I wonder if it makes sense to exercise more corner cases here:

*
1. Select one node
2. Call the second time, make sure than the second node is selected, and it is different than the first one
3. Call the third time, make sure that no additional nodes have been selected.

*
1. Add only a single node to the InmemoryNodeManager
2. Call once, make sure it is selected
3. Call the second time, make sure the output is the same.
4. Add one more node to the InmemoryNodeManager
5. Call the third time, make sure it has a new node selected
6. Add one more node to the InmemoryNodeManager
7. Call for the forth time, make sure the latest added node is not selected, as the limit is reached.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java (hunk 14 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java]: Agreed with @arhimondr . Calling it per each `computeAssignments` call might be expensive. 

Also, we usually have new lines for stream API call:
```java
existingTasks.stream()
      .map(remoteTask -> nodeMap.getNodesByNodeId().get(remoteTask.getNodeId()))
      .collect(toSet());
```

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java (hunk 10 above)
Comment (wenleix) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: @arhimondr , @cemcayiroglu : If we are OK to have `maxTasksPerStage` when creating the `NodeSelector`, do we want to just pre-compute a limited set of nodes when `NodeSelector` is created? 

Thus the only reason why need to consult `nodeManager` during `computeAssignments` is when the query is submitted at cluster initialization time, which is not a common case.

**Comment by wenleix** [comment]
@arhimondr : This is off-topic of this PR. But here is the ancient discussion about `ResettableRandomizedIterator`: 
https://github.com/prestodb/presto/pull/12373#discussion_r260892611 

I see your point... but anyhow, it's not a "common iterator" (the long name also suggests it 😉  )

**Comment by cemcayiroglu** [comment]
@wenleix thanks for your feedback! We are recomputing the nodes again since the nodemap gets refreshed every 5 seconds. New worker nodes can join while we are scheduling the splits. We can cache the nodes if they reach the limit but this also makes the code more complicated and stateful. I think we can keep it as is (after adding @arhimondr suggestions. I would like to start refactoring the interfaces after this PR :) The interface names dont make sense.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java (hunk 14 above)
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java]: nit: `toImmutableSet`

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java [hunk 16]
```diff
@@ -408,6 +408,8 @@ private static URI getBufferLocation(RemoteTask remoteTask, OutputBufferId rootB
 
         Optional<int[]> bucketToPartition;
         PartitioningHandle partitioningHandle = plan.getFragment().getPartitioning();
+        int maxTasksPerStage = getMaxTasksPerStage(session);
```
Comment (highker) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: Do we wanna do some sanity check on this value?

File: presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java (hunk 15 above)
Comment (cemcayiroglu) [reply presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java]: @arhimondr I couldnt understand the first test case. select one node twice?

Comment (arhimondr) [reply presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java]: e.g.: provide only a single split. So only a single node of three is selected at the first step.

**Comment by wenleix** [comment]
@cemcayiroglu 
> @wenleix thanks for your feedback! We are recomputing the nodes again since the nodemap gets refreshed every 5 seconds. New worker nodes can join while we are scheduling the splits. We can cache the nodes if they reach the limit but this also makes the code more complicated and stateful. I think we can keep it as is (after adding @arhimondr suggestions. I would like to start refactoring the interfaces after this PR :) The interface names dont make sense.

Correct, what I mean is we only do recomputing when we don't have enough nodes for schedule. I think @arhimondr  suggests some similar ideas. But I am open to wait for the refactor/cleanup given the current interface is a bit messy already. 

Another idea is can we skip this recomputing when `limit` is large enough? -- I remember this `max-task-per-stage` is `2^31-1` by default. So in the default case this computing doesn't need to kicked in :)

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java (hunk 10 above)
Comment (arhimondr) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: > Thus the only reason why need to consult nodeManager during computeAssignments is when the query is submitted at cluster initialization time, which is not a common case.

I'm not sure how important that in practice, but i think it is better to preserve the existing behaviour.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java [hunk 17]
```diff
@@ -68,19 +68,21 @@
     private final int maxPendingSplitsPerTask;
     private final NodeTaskMap nodeTaskMap;
     private final boolean useNetworkTopology;
+    private final long nodeMapRefreshInNanos;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java]: Please use `io.airlift.units.Duration` instead.

See the 9th bullet point here: https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java [hunk 18]
```diff
@@ -68,19 +68,21 @@
...
     }
 
     public NodeScheduler(
             NetworkLocationCache networkLocationCache,
             NetworkTopology networkTopology,
             InternalNodeManager nodeManager,
             NodeSchedulerConfig config,
-            NodeTaskMap nodeTaskMap)
+            NodeTaskMap nodeTaskMap,
+            long nodeMapRefreshInNanos)
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java]: Accept `Duration` here

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java [hunk 19]
```diff
@@ -159,8 +169,8 @@ public NodeSelector createNodeSelector(ConnectorId connectorId)
                 }
             }
 
-            return new NodeMap(byHostAndPort.build(), byHost.build(), workersByNetworkPath.build(), coordinatorNodeIds);
-        }, 5, TimeUnit.SECONDS);
+            return new NodeMap(byNodeId.build(), byHostAndPort.build(), byHost.build(), workersByNetworkPath.build(), coordinatorNodeIds);
+        }, nodeMapRefreshInNanos, TimeUnit.NANOSECONDS);
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java]: The `memoizeWithExpiration` supplier doesn't allow 0 =\

I recommend to have a condition here. If the duration is 0, then create a regular supplier. Otherwise the `memoizeWithExpiration` one.

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java [hunk 20]
```diff
@@ -168,6 +172,23 @@ else if (!splitWaitingForAnyNode) {
...
+        List<InternalNode> existingNodes = existingTasks.stream()
+                .map(remoteTask -> nodeMap.getNodesByNodeId().get(remoteTask.getNodeId()))
+                .collect(toList());
+
+        int alreadySelectedNodeCount = existingTasks.size();
+        int nodeCount = nodeMap.getNodesByNodeId().size();
+
+        if (alreadySelectedNodeCount < limit && alreadySelectedNodeCount < nodeCount) {
+            List<InternalNode> moreNodes =
+                    selectNodes(limit - alreadySelectedNodeCount, randomizedNodes(nodeMap, includeCoordinator, Sets.newHashSet(existingNodes)));
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/SimpleNodeSelector.java]: We prefer immutable collections whenever possible. Please use `ImmutableSet.copyOf()`.

See the second bullet point here: https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development

File: presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java [hunk 21]
```diff
@@ -424,6 +425,97 @@ public void testSplitCount()
...
+        NodeTaskMap nodeTaskMap = new NodeTaskMap(finalizerService);
+        TestingTransactionHandle transactionHandle = TestingTransactionHandle.create();
+        NodeSchedulerConfig nodeSchedulerConfig = new NodeSchedulerConfig()
+                .setMaxSplitsPerNode(20)
+                .setIncludeCoordinator(false)
+                .setMaxPendingSplitsPerTask(10);
+
+        LegacyNetworkTopology networkTopology = new LegacyNetworkTopology();
+        // refresh interval is 1 nanosecond
+        NodeScheduler nodeScheduler = new NodeScheduler(new NetworkLocationCache(networkTopology), networkTopology, nodeManager, nodeSchedulerConfig, nodeTaskMap, 1);
```
Comment (arhimondr) [inline presto-main/src/test/java/com/facebook/presto/execution/TestNodeScheduler.java]: The `memoizeWithExpiration` supplier uses the `System.nanoTime()` to check if the memoize duration has passed.

From the `System.nanoTime()` javadoc:

> This method provides nanosecond precision, but not necessarily nanosecond resolution

It means that the `System.nanoTime()` may not return the updated value even if non zero number of nanoseconds have passed.

Although it is rather unlikely for the calls to take less then 1 ns, it still possible for `System.nanoTime()` to do not provide enough nanotime resolution at some platforms. 

This behaviour can result in a flakiness of this test. Thus i recommend to have ` ... condition here. If the duration is 0, then create a regular supplier. Otherwise the memoizeWithExpiration one ... `

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java (hunk 10 above)
Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: Looks like this part of the code is super messy. nodescheduler and nodeselector is calling each other. no one knows who is doing what.  I am creating an issue to address this. please look: https://en.wikipedia.org/wiki/Spaghetti_code

Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: https://github.com/prestodb/presto/issues/13553

Comment (cemcayiroglu) [reply presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java]: https://github.com/prestodb/presto/issues/13553

File: presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java [hunk 22]
```diff
@@ -68,19 +69,21 @@
     private final int maxPendingSplitsPerTask;
     private final NodeTaskMap nodeTaskMap;
     private final boolean useNetworkTopology;
+    private final Duration nodeMapRefreshTime;
```
Comment (arhimondr) [inline presto-main/src/main/java/com/facebook/presto/execution/scheduler/NodeScheduler.java]: nodeMapRefreshInterval

**Comment by arhimondr** [comment]
@wenleix @highker It looks like all the comments are addressed. I'm going to merge it soon. Let me know if you got any additional comments.

**Comment by cemcayiroglu** [comment]
@highker, @wenleix, @arhimondr do you all have any additional comments?
